pkg/controlplane/apiserver/options/options.go

```diff
@@ -262,6 +269,23 @@ func (o *Options) Complete(alternateDNS []string, alternateIPs []net.IP) (Comple
 			if err != nil {
 				return CompletedOptions{}, fmt.Errorf("failed to build token generator: %w", err)
 			}
+		} else if completed.ServiceAccountSigningEndpoint != "" {
+			plugin, cache, err := plugin.New(completed.Authentication.ServiceAccounts.Issuers[0], completed.ServiceAccountSigningEndpoint)
+			if err != nil {
+				return CompletedOptions{}, fmt.Errorf("while setting up external-jwt-signer: %w", err)
+			}
+			metadata, err := plugin.GetServiceMetadata(context.Background())
+			if err != nil {
+				return CompletedOptions{}, fmt.Errorf("while setting up external-jwt-signer: %w", err)
+			}
+			if metadata.MaxTokenExpirationSeconds < 600 {
+				return CompletedOptions{}, fmt.Errorf("Max token life supported by external-jwt-signer (%ds) is less than accaptable (min 600s)", metadata.MaxTokenExpirationSeconds)
+			}
+			if metadata.MaxTokenExpirationSeconds < int(completed.Authentication.ServiceAccounts.MaxExpiration.Seconds()) {
```

Comment:
this should only error if they explicitly set `completed.Authentication.ServiceAccounts.MaxExpiration` (override via the flag) in a way that conflicted with the external signer 

see the OIDC example for how we can hook in a function to be able to tell if a particular flag was changed

```
		o.OIDC.areFlagsConfigured = func() bool {
			return fs.Changed(oidcIssuerURLFlag) ||
```

if they accepted the kube-apiserver defaults and the external signer max duration is less, just use the external signer's max duration (override `completed.Authentication.ServiceAccounts.MaxExpiration`)

That means we need to move the `// verify and adjust ServiceAccountTokenMaxExpiration` block *after* this block so the overridden value is used

It *also* means the warning when `completed.Authentication.ServiceAccounts.MaxExpiration < serviceaccount.ExpirationExtensionSeconds*time.Second` needs to be adjusted when using an external signer, because the external signer's max expiration will rule

it *also* means when using an external signer and extended duration tokens, the max expiration is a hard limit, so this code in the token issuing path needs to know when the max expiration is a hard limit and cap it:

```go
	if r.extendExpiration && pod != nil && req.Spec.ExpirationSeconds == token.WarnOnlyBoundTokenExpirationSeconds && r.isKubeAudiences(req.Spec.Audiences) {
		warnAfter = exp
		exp = token.ExpirationExtensionSeconds
	}
```

Reply:
Done.

Can you take another look to confirm if you agree with what I did?

Reply:
Resolving as we went over it in the latest pass.